Python/Product/Analysis/ProjectEntry.cs

```diff
@@ -342,15 +320,118 @@ where fileName.StartsWith(pathPrefix)
             _aggregates.Add(into);
         }
+
+        public IEnumerable<int> DocumentParts => _buffers.Keys.AsLockedEnumerable(_buffers);
+
+        public Stream ReadDocumentBytes(int part, out int version) {
+            lock (_buffers) {
+                if (_buffers[part].Version >= 0) {
+                    version = _buffers[part].Version;
+                    return EncodeToStream(_buffers[part].Text, Encoding.UTF8);
+                } else if (part == 0) {
+                    var s = PathUtils.OpenWithRetry(FilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                    version = s == null ? -1 : 0;
+                    return s;
+                }
+                version = -1;
+                return null;
+            }
+        }
+
+        internal static Stream EncodeToStream(StringBuilder text, Encoding encoding, int chunkSize = 4096) {
+            byte[] bytes, preamble;
+            MemoryStream ms;
+
+            if (text.Length < chunkSize) {
+                preamble = encoding.GetPreamble() ?? Array.Empty<byte>();
+                bytes = encoding.GetBytes(text.ToString());
+                if (preamble.Any()) {
+                    ms = new MemoryStream(preamble.Length + bytes.Length);
+                    ms.Write(preamble, 0, preamble.Length);
+                    ms.Write(bytes, 0, bytes.Length);
+                    ms.Seek(0, SeekOrigin.Begin);
+                    return ms;
+                }
+                return new MemoryStream(bytes);
+            }
+
+            // Estimate 1 byte per character (or 2 bytes for UTF-16) for initial allocation
+            ms = new MemoryStream(Encoding.Unicode.Equals(encoding) ? text.Length * 2 : text.Length);
+            var enc = encoding.GetEncoder();
+            var chars = new char[chunkSize];
+
+            preamble = encoding.GetPreamble() ?? Array.Empty<byte>();
+            ms.Write(preamble, 0, preamble.Length);
+
+            bytes = new byte[encoding.GetMaxByteCount(chunkSize)];
+            for (int i = 0; i < text.Length;) {
+                bool flush = true;
+                int len = text.Length - i;
+                if (len > chars.Length) {
+                    flush = false;
+                    len = chars.Length;
+                }
+                text.CopyTo(i, chars, 0, len);
+                enc.Convert(chars, 0, len, bytes, 0, bytes.Length, flush, out int charCount, out int bytesUsed, out _);
+                i += charCount;
+                ms.Write(bytes, 0, bytesUsed);
+            }
+            ms.Seek(0, SeekOrigin.Begin);
+            return ms;
+        }
+
+        public TextReader ReadDocument(int part, out int version) {
+            var stream = ReadDocumentBytes(part, out version);
+            if (stream == null) {
+                version = -1;
+                return null;
+            }
+            try {
+                var sr = Parser.ReadStreamWithEncoding(stream, ProjectState.LanguageVersion);
```

Comment:
`ReadStreamWithEncoding` wraps `stream` with `PartiallyReadStream` and then `stream` is set to `null`, so when `TextReader` is disposed, underlying stream will stay alive.